web/packages/teleport/src/Roles/RoleEditor/StandardEditor/MetadataSection.test.tsx

```diff
@@ -0,0 +1,156 @@
+        return {
+          kind: 'role',
+          id: '',
+          name,
+          content: '',
+        };
+      }
+      throw new ApiError({
+        message: `role ${name} is not found`,
+        response: { status: 404 } as Response,
+      });
+    });
+});
+
+afterEach(() => {
+  jest.restoreAllMocks();
+});
+
+const setup = () => {
+  const modelRef = jest.fn();
+  const ctx = createTeleportContext();
+  let validator: Validator;
+  let dispatch: StandardModelDispatcher;
+  render(
+    <TeleportContextProvider ctx={ctx}>
+      <StatefulSectionWithDispatch<MetadataModel, MetadataValidationResult>
+        selector={m => m.roleModel.metadata}
+        validationSelector={m => m.validationResult.metadata}
+        component={MetadataSection}
+        validatorRef={v => {
+          validator = v;
+        }}
+        modelRef={modelRef}
+        dispatchRef={d => {
+          dispatch = d;
+        }}
+      />
+    </TeleportContextProvider>
+  );
+  return { modelRef, dispatch, validator };
+};
+
+test('basic editing', async () => {
+  const user = userEvent.setup();
+  const { modelRef } = setup();
+  await user.clear(screen.getByLabelText('Role Name *'));
+  await user.type(screen.getByLabelText('Role Name *'), 'some-name');
+  await user.type(screen.getByLabelText('Description'), 'some-description');
+  await user.click(screen.getByRole('button', { name: 'Add a Label' }));
+  await user.type(screen.getByPlaceholderText('label key'), 'foo');
+  await user.type(screen.getByPlaceholderText('label value'), 'bar');
+  await selectEvent.select(screen.getByLabelText('Version'), 'v6');
+  expect(modelRef).toHaveBeenLastCalledWith({
+    name: 'some-name',
+    nameCollision: false,
+    description: 'some-description',
+    labels: [{ name: 'foo', value: 'bar' }],
+    version: { label: 'v6', value: 'v6' },
+  } as MetadataModel);
+});
+
+test('basic validation', async () => {
+  const user = userEvent.setup();
+  const { validator } = setup();
+  await user.clear(screen.getByLabelText('Role Name *'));
+  await user.click(screen.getByRole('button', { name: 'Add a Label' }));
+  act(() => validator.validate());
+
+  expect(screen.getByLabelText('Role Name *')).toHaveAccessibleDescription(
+    'Role name is required'
+  );
+  expect(screen.getByPlaceholderText('label key')).toHaveAccessibleDescription(
+    'required'
+  );
+});
+
+// We are testing debounced logic and we don't want the test to wait until the
+// timer fires, so we are using fake timers. Because of that, we wrap the test
+// with a custom pair of before/after routines.
+describe('asynchronous validation', () => {
+  beforeEach(() => {
+    jest.useFakeTimers();
+  });
+
+  afterEach(() => {
+    jest.runOnlyPendingTimers();
+    jest.useRealTimers();
+  });
+
+  test('checking for existing roles', async () => {
+    // Required for userEvent to cooperate nicely with fake timers.
+    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
```

Comment:
Oh lol maybe that's what I missed when I was writing my tests with fake timers, idk.